docs/protocol.md

```diff
@@ -0,0 +1,364 @@
+# HTTP protocol for RAG chat apps
+
+We are standardizing on a common HTTP protocol across RAG chat app solutions and tools,
+to make them more compatible with each other. By agreeing on a common protocol, developers
+can swap different components, like using the Python backend from this repository
+with a Web Components frontend from [azure-search-openai-javascript](https://github.com/Azure-Samples/azure-search-openai-javascript).
+
+This protocol is inspired by the [OpenAI ChatCompletion API](https://platform.openai.com/docs/guides/text-generation/chat-completions-api),
+but contains additional fields required for a RAG chat app.
+
+## HTTP requests to RAG chat app endpoints
+
+An HTTP request to `/chat` or `/ask` can contain these properties, in JSON:
+
+* `"messages"`: A list of messages, each containing "content" and "role", where "role" may be "system", "assistant", or "user"
+* `"stream"`: A boolean indicating whether the response should be streamed or not.
+* `"context"`: An object containing any additional context about the request. This app only sets the "overrides" property, which is an object with additional settings for the request, like how many search results to return.
+* `"session_state"`: An object containing the "memory" for the chat app, such as a user ID. Not currently used by this app.
+
+Here's an example JSON request:
+
+```json
+{
+    "messages": [
+        {
+            "content": "What is included in my Northwind Health Plus plan that is not in standard?",
+            "role": "user"
+        }
+    ],
+    "stream": false,
+    "context": {
```

Comment:
Context is how servers extend the protocol? Are clients that are context-unaware guaranteed to work with the server?

Reply:
Yep, context is basically the customization mechanism for both requests/responses. I try to treat context as optional where possible, to increase compatibility - i.e. in theory a non-RAG chat app client could interface with our backend and still work, it just wouldn't show citations.

Reply:
We've extended this protocol with Authorization headers which aren't captured here (pure json request / response). Should we document this extension in the protocol spec?

Reply:
Yes, documented here: https://github.com/Azure-Samples/ai-chat-app-protocol/pull/1/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R30
Let me know if that's described correctly.